packages/components/src/Datalist/Datalist.component.js

```diff
@@ -14,111 +15,295 @@ export function escapeRegexCharacters(str) {
-			newState.titleMapping = this.buildTitleMapping(titleMap);
-			stateCallback = () => {
-				if (this.state.suggestions) {
-					const filter = value !== this.state.value ? this.state.value : null;
-					this.updateSuggestions(filter);
-				}
-			};
+	if (displayMode === DISPLAY.ALL || !filterValue) {
+		return titleMap;
+	}
+
+	// building multiSection items or single section items
+	const escapedValue = escapeRegexCharacters(filterValue.trim());
+	const regex = new RegExp(escapedValue, 'i');
+	if (multiSection) {
+		return titleMap
+			.map(group => ({
+				...group,
+				suggestions: filterValue
+					? group.suggestions.filter(item => regex.test(item.name))
+					: group.suggestions,
+			}))
+			.filter(group => group.suggestions.length > 0);
+	}
+
+	// only one group so items are inline
+	return titleMap.filter(itemValue => regex.test(itemValue.name));
+}
+
+function findEntry(titleMap, attributeName, attributeValue = '') {
+	if (!titleMap) {
+		return null;
+	}
+
+	for (let index = 0; index < titleMap.length; index += 1) {
+		const entry = titleMap[index];
+
+		if (entry.suggestions) {
+			// entry is a category, it has suggestions that are real entries {name, value}
+			const categorySuggestion = entry.suggestions.find(
+				item =>
+					!item.disabled && item[attributeName].toLowerCase() === attributeValue.toLowerCase(),
+			);
+			if (categorySuggestion) {
+				return categorySuggestion;
+			}
+		} else if (
+			!entry.disabled &&
+			entry[attributeName].toLowerCase() === attributeValue.toLowerCase()
+		) {
+			// entry is {name, value}
+			return entry;
 		}
-		this.setState(newState, stateCallback);
 	}
 
+	return null;
+}
+
+function getEntryFromName(titleMap, name, restricted) {
+	const entry = findEntry(titleMap, 'name', name);
+	if (entry) {
+		return entry;
+	}
+
+	return restricted ? undefined : { name, value: name };
+}
+
+function getEntryFromValue(titleMap, value, restricted) {
+	const entry = findEntry(titleMap, 'value', value);
+	if (entry) {
+		return entry;
+	}
+
+	return restricted ? undefined : { name: value, value };
+}
+
+function getEntry(titleMap, nameOrValue, restricted) {
+	const entry =
+		findEntry(titleMap, 'name', nameOrValue) || findEntry(titleMap, 'value', nameOrValue);
+	if (entry) {
+		return entry;
+	}
+
+	return restricted ? undefined : { name: nameOrValue, value: nameOrValue };
+}
+
+function Datalist(props) {
+	// Current persisted value
+	// In case of simple values, this is a string
+	// In case of titleMap, it's an object { name: "display value", value: "technical value" }
+	const [value, setValue] = useState();
+
+	// suggestions: filter value, display flag, current hover selection
+	const [filterValue, setFilterValue] = useState('');
+	const [displaySuggestion, setDisplaySuggestion] = useState(DISPLAY.NONE);
+	const [selection, setSelection] = useState({
+		focusedItemIndex: undefined,
+		focusedSectionIndex: undefined,
+	});
+
+	// suggestions computation
+	const suggestions = useMemo(
```

Comment:
```suggestion
	const suggestions = useMemo(
```
Is there an observable benefit to useMemo here?

Reply:
hmm indeed, it changes at every change in fact.
With you comment I even try to think about possible optimisation like decoupling the hide/display from the suggestion computation, but when we hide the suggestions, the next open needs either
-  to have the whole liste, not filtered one, in case of newly focus/click, or key down/up
- to update the filtered suggestions in case we already are in input but closes suggestions and type in it again

Reply:
Removed it